pkg/datamover/dataupload_delete_action.go

```diff
@@ -0,0 +1,86 @@
+package datamover
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1api "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	velerov1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	velerov2alpha1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v2alpha1"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+	"github.com/vmware-tanzu/velero/pkg/repository"
+)
+
+type DataUploadDeleteAction struct {
+	logger logrus.FieldLogger
+	client client.Client
+}
+
+func (d *DataUploadDeleteAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"datauploads.velero.io"},
+	}, nil
+}
+
+func (d *DataUploadDeleteAction) Execute(input *velero.DeleteItemActionExecuteInput) error {
+	d.logger.Infof("Executing DataUploadDeleteAction")
+	du := &velerov2alpha1.DataUpload{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(input.Item.UnstructuredContent(), &du); err != nil {
+		return errors.WithStack(errors.Wrapf(err, "failed to convert input.Item from unstructured"))
+	}
+	cm := genConfigmap(input.Backup, *du)
+	if cm == nil {
+		// will not fail the backup deletion
+		return nil
+	}
+	err := d.client.Create(context.Background(), cm)
+	if err != nil {
+		return errors.WithStack(errors.Wrapf(err, "failed to create the configmap for DataUpload %s/%s", du.Namespace, du.Name))
+	}
+	return nil
```

Comment:
One more thing, if the DU CR exists in the current cluster, we should also delete the DU CR.
Just a simple delete is fine:

- if the DU is not in progress, the controller will handle the cleanup appropriately. cc @qiuming-best 
- otherwise,  a CM is created and we can use the CM to track the snapshot info regardless whether the snapshot can be deleted successfully later

Reply:
Oh yes, sure, updated.